ns2-stat/src/lib.rs

```diff
@@ -74,65 +78,130 @@ pub struct NS2Stats {
-                WinningTeam::Aliens => {
-                    map_entry.alien_wins += 1;
-                    alien_wins += 1;
+                None => {
+                    self.users.insert(key, value);
                 }
-                WinningTeam::None => {}
             }
+        }
 
-            if game.round_info.round_date > latest_game {
-                latest_game = game.round_info.round_date;
+        for (key, value) in other.maps {
+            match self.maps.get_mut(&key) {
+                Some(t) => {
+                    t.merge(value);
+                }
+                None => {
+                    self.maps.insert(key, value);
+                }
             }
-            total_games += 1;
         }
+    }
+}
 
-        for user in users.values_mut() {
-            user.kd = user.kills as f32 / user.deaths as f32;
-            user.kda = (user.kills + user.assists) as f32 / user.deaths as f32;
-        }
+impl FromIterator<NS2Stats> for Option<NS2Stats> {
+    fn from_iter<T: IntoIterator<Item = NS2Stats>>(iter: T) -> Self {
+        iter.into_iter().reduce(|mut acc, item| {
+            acc.merge(item);
+            acc
+        })
+    }
+}
 
-        Self {
-            latest_game,
+impl From<GameStats> for NS2Stats {
+    fn from(gs: GameStats) -> Self {
+        (&gs).into()
+    }
+}
+
+impl From<&GameStats> for NS2Stats {
+    fn from(game: &GameStats) -> Self {
+        use std::collections::hash_map::Entry;
+        let mut stats = Self::default();
+
+        let Self {
             users,
             maps,
-            total_games,
-            marine_wins,
             alien_wins,
+            latest_game,
+            marine_wins,
+            total_games,
+        } = &mut stats;
+
+        for player_stat in game.player_stats.values() {
+            let user = match users.entry(player_stat.player_name.clone()) {
+                Entry::Occupied(o) => o.into_mut(),
+                Entry::Vacant(v) => v.insert(User::default()),
+            };
+
+            user.total_games += 1;
+
+            for stats in [&player_stat.marines, &player_stat.aliens] {
+                user.kills += stats.kills;
+                user.assists += stats.assists;
+                user.deaths += stats.deaths;
+            }
+        }
+
+        let map_entry = match maps.entry(game.round_info.map_name.clone()) {
+            Entry::Occupied(o) => o.into_mut(),
+            Entry::Vacant(v) => v.insert(Map::default()),
+        };
+
+        map_entry.total_games += 1;
+        match game.round_info.winning_team {
+            WinningTeam::Marines => {
+                map_entry.marine_wins += 1;
+                *marine_wins += 1;
+            }
+            WinningTeam::Aliens => {
+                map_entry.alien_wins += 1;
+                *alien_wins += 1;
+            }
+            WinningTeam::None => {}
         }
+
+        *latest_game = game.round_info.round_date;
+        *total_games += 1;
+        stats
```

Comment:
```suggestion
        let mut users = HashMap::new();
        for player_stat in game.player_stats.values() {
            let user = match users.entry(player_stat.player_name.clone()) {
                Entry::Occupied(o) => o.into_mut(),
                Entry::Vacant(v) => v.insert(User::default()),
            };

            user.total_games += 1;

            for stats in [&player_stat.marines, &player_stat.aliens] {
                user.kills += stats.kills;
                user.assists += stats.assists;
                user.deaths += stats.deaths;
            }
        }

        let mut marine_wins = 0;
        let mut alien_wins = 0;
        let mut map = Map {
            total_games: 1,
            ..Map::default()
        };
        match game.round_info.winning_team {
            WinningTeam::Marines => {
                map.marine_wins = 1;
                marine_wins = 1;
            }
            WinningTeam::Aliens => {
                map.alien_wins = 1;
                alien_wins = 1;
            }
            WinningTeam::None => {}
        }

        Self {
            latest_game: game.round_info.round_date,
            users,
            maps: HashMap::from([(game.round_info.map_name.clone(), map)]),
            total_games: 1,
            marine_wins,
            alien_wins,
        }
```
(This is supposed to replace the whole function body, but suggestions don't work for deleted lines.)